### ISSUE TITLE
OOM when reading Parquet file

### ISSUE THREAD

**Issue by alexey-milovidov** [duckdb/duckdb#3969]
#### What happens?
It is using all available memory and is terminated by OOM.

#### To Reproduce

Allocate a machine with 32 GB RAM, like c6a.4xlarge on AWS, with Ubuntu 22.04.
ssh into that machine.
Run the following commands:

```
sudo apt-get update
sudo apt-get install python3-pip
pip install duckdb
wget 'https://datasets.clickhouse.com/hits_compatible/hits.parquet'
```

Create the following run.py file:
```
#!/usr/bin/env python3

import duckdb
import timeit

con = duckdb.connect(database='my-db.duckdb', read_only=False)

print("Will load the data")

start = timeit.timeit()
con.execute("CREATE TABLE hits AS SELECT * FROM parquet_scan('hits.parquet')")
end = timeit.timeit()
print(end - start)
```

Make it executable:
```
chmod +x run.py
```

Run it:
```
./run.py
```

Wait around 10 minutes... 

```
Will load the data
Killed
```

#### Environment (please complete the following information):
 - OS: Ubuntu 22.04
 - DuckDB Version: 0.4.0
 - DuckDB Client: Python

#### Identity Disclosure:
 - Full Name: Alexey Milovidov
 - Affiliation: ClickHouse, Inc

With OOM it cannot qualify in the ClickHouse benchmark.

#### Before Submitting

- [x] **Have you tried this on the latest `master` branch?** No.
* **Python**: `pip install duckdb --upgrade --pre` It installs the same version 0.4.0.
* **R**: I don't use R.
* **Other Platforms**: I don't use other platforms.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?** Yes.

**Comment by alexey-milovidov** [comment]
The table contains 99 997 497 rows.

**Comment by alexey-milovidov** [comment]
I tried to add the following lines to the Python script:

```
con.execute("PRAGMA memory_limit='{}'".format(psutil.virtual_memory().total))
con.execute("PRAGMA threads={}".format(psutil.cpu_count(logical=False)))
```

But it did not help.

**Comment by alexey-milovidov** [comment]
I tried to load from the CLI but it also fails with OOM:

```
#!/usr/bin/expect -f

match_max 100000
set timeout 10000

spawn ./duckdb
expect "D "

send -- "PRAGMA threads = 8;\r"
expect "D "

send -- "PRAGMA temp_directory='duckdb.tmp';\r"
expect "D "

send -- ".timer on\r"
expect "D "

send -- "CREATE TABLE hits AS SELECT * FROM parquet_scan('hits.parquet');\r"
expect "D "
```

**Comment by alexey-milovidov** [comment]
Changing `parquet_scan` to `read_parquet` does not help, although the difference is unclear.

**Comment by Alex-Monahan** [comment]
Hello!

Just a random idea, but would it help to create the table first and then insert into it? Create table as select * from ... where 1=0, etc.

**Comment by alexey-milovidov** [comment]
I tried, but it did not help.
Also tried to switch to .csv.

**Comment by Mytherin** [comment]
Thanks for the report! I will investigate more after the weekend.

As a work-around for now, perhaps try either (1) setting a lower memory limit explicitly (e.g. `SET memory_limit='8GB'`) or (2) splitting up the file load into multiple calls using e.g. `INSERT INTO tbl SELECT * FROM hits.parquet LIMIT ... OFFSET ...`. Neither will be particularly fast but at least it should work.

**Comment by Mytherin** [comment]
On the topic of the clickhouse benchmarks, please make sure to use the `approx_count_distinct` function instead of `COUNT(DISTINCT)` for better performance on approx distinct counts :)

Is the code that you are using to benchmark DuckDB available somewhere publicly, or could we perhaps have a look at it otherwise?

**Comment by hannes** [comment]
(Hi, Alexey), I tried this, it went up to 117 GB on a machine with 126 GB RAM when not setting a memory limit, but it completed. When setting the memory limit to 10 GB, it went up to 24 GB but indeed completed successfully (albeit much slower). 

So indeed I think we need to have a look for the source of those additional allocations.

**Comment by alexey-milovidov** [comment]
Ok, I will set the memory limit to 8GB.

I'm currently preparing the benchmark, it is in work in progress: https://github.com/ClickHouse/ClickHouse/pull/38437
The idea is to make it easy reproducible (like running a single small script) and also compatible with most SQL DBMS.
So, I've removed all custom aggregate functions and data types.

**Comment by Mytherin** [comment]
Sounds great! Let me know if the fix works and if you run into anything else.

**Comment by alexey-milovidov** [comment]
Strange, but it does not work. I've set memory limit to the `psutil.virtual_memory().total / 4`.

If I load Parquet it does not succeed due to unsupported cast:
```
Unimplemented type for cast (BIGINT -> TIMESTAMP)
```

If I load CSV it still failed with OOM.

I will try to use 4GB.

**Comment by alexey-milovidov** [comment]
I will try to split the file into parts.

**Comment by alexey-milovidov** [comment]
Now I have different error:

```
ubuntu@ip-172-31-4-239:~$ ./run.py 
Will load the data
4216.5390389899985
SELECT COUNT(*) FROM hits;

0.05523904401343316
0.003909286999260075
0.0038267960044322535
SELECT COUNT(*) FROM hits WHERE AdvEngineID != 0;

0.1603328409983078
0.06337512300524395
0.05665142499492504
SELECT SUM(AdvEngineID), COUNT(*), AVG(ResolutionWidth) FROM hits;

0.238524653002969
0.08392444600758608
0.07758511300198734
SELECT AVG(UserID) FROM hits;

1.1524075900088064
0.10763906499778386
0.09490979700058233
SELECT COUNT(DISTINCT UserID) FROM hits;

9.184395752003184
8.429843456004164
9.689035353003419
SELECT COUNT(DISTINCT SearchPhrase) FROM hits;

Segmentation fault (core dumped)
```

**Comment by alexey-milovidov** [comment]
If I restart Python script after every query (and reopen the database), it succeeded.
Most likely previous queries have corrupted the memory.

**Comment by hannes** [comment]
That's also something for us to look into. But to come back on a point Mark made above, I see you are running COUNT DISTINCT there, which are approximate in ClickHouse AFAIK but are not approximate (ie exact) in DuckDB. Could you please make sure DuckDB's approx_count_distinct for the comparison? Thanks!

**Comment by alexey-milovidov** [comment]
No, COUNT(DISTINCT ...) is exact in ClickHouse.
ClickHouse also has an approximate count distinct, but under different names.

**Comment by hannes** [comment]
Thanks for the clarification, great to hear the default is the exact one.

**Comment by alexey-milovidov** [comment]
Thank you!

ClickHouse is using jemalloc since [Aug 2018](https://github.com/ClickHouse/ClickHouse/pull/2773).
We have found and fixed multiple bugs in jemalloc, and continue to find them, but generally, it works ok.

And glibc's malloc is something that no reasonable software should ever use.

**Comment by alexey-milovidov** [comment]
@hannes I've included DuckDB into https://github.com/ClickHouse/ClickBench
Hope it will be interesting.

**Comment by Mytherin** [comment]
This should be fully resolved now that jemalloc is merged (#4971) and we stream data to disk optimistically (#4996). Feel free to re-open if you are still hitting the issue.
